src/vision/status_quo/barbara_battles_buffered_streams.md

```diff
@@ -0,0 +1,216 @@
+
+### **How would this story have played out differently for the other characters?**
+
+Woe be unto them! Identifying and fixing this bug required a lot of fluency with Rust and the async model. Alan in particular was probably relying on his understanding of async-await from other languages, which works very differently. In those languages, every async function is enqueued automatically for independent execution, so hazards like this do not arise (though this comes at a performance cost).
+
+### Besides timeouts for clients, what else could go wrong?
+
+The original bug report mentioned the possibility of deadlock:
+
+> When using an async friendly semaphore (like Tokio provides), you can deadlock yourself by having the tasks that are waiting in the `FuturesUnordered` owning all the semaphores, while having an item in a `.for_each()` block after `buffer_unordered()` requiring a semaphore.
+
+### Is there any way for Barbara to both produce and process work items simultaneously?
+
+Yes, in this case, she could've. For example, she might have written
+
+```rust
+async fn do_work(database: &Database) {
+    let work = do_select(database, FIND_WORK_QUERY).await?;
+
+    stream::iter(work)
+        .map(|item| async move {
+            let work_item = do_select(database, work_from_item(item)).await;
+            process_work_item(database, work_item).await;
+        })
+        .buffered(5)
+        .for_each(|()| std::future::ready(()))
+        .await;
+}
+```
+
+This would mean however that limit the number of work items being processed to 5. This may be what she wanted, but perhaps not. It may also be that similar problems come up in scenarios where Barbara can't so easily alter the call to `do_select` to "append" a call to `process_work_item` (for example, maybe she is in some function that takes a [`FuturesUnordered`] or `Stream` as argument).
+
+### Is there any way for Barbara to both produce and process work items simultaneously, without the buffering and so forth?
+
+Yes, she might use a loop with a `select!`. This would ensure that she is processing *both* the stream that produces work items and the [`FuturesUnordered`] that consumes them:
+
+```rust
+async fn do_work(database: &Database) {
+    let work = do_select(database, FIND_WORK_QUERY).await?;
+
+    let selects = stream::iter(work)
+        .map(|item| do_select(database, work_from_item(item)))
+        .buffered(5)
+        .fuse();
+    tokio::pin!(selects);
+
+    let mut results = FuturesUnordered::new();
+
+    loop {
+        tokio::select! {
+            Some(work_item) = selects.next() => {
+                results.push(process_work_item(database, work_item));
+            },
+            Some(()) = results.next() => { /* do nothing */ },
+            else => break,
+        }
+    }
+}
+```
+
+Note that doing so is producing code that looks quite a bit different than where she started, though. :(
```

Comment:
This also behaves very differently - we can have a queue of tens of thousands of items that `do_select` grabs from, and we don't want to pull too many items out of the queue (because then we'd have to re-queue them on shutdown).

The intent of `buffered(5)` here was to allow us to grab 5 work items from the queue at most, so that other hosts could pull out work items and share the load when there's a spike. It doesn't work like that, unfortunately :-(

Reply:
I incorporated this